Erster Versuch für das Versenden einer E-Mail an eine Kontaktperson

aletutto opened the pull request:
Habe mal einen ersten Versuch gestartet, um eine Mail an eine Kontaktperson zu verschicken.
Dafür habe ich uns eine Mail-Adresse bei Google angelegt: fae.nachrichtensystem@gmail.com.

### Weboberfläche
Für den Anfang habe ich auch mal eine sehr (wirklich sehr) simple Web-Oberfläche dafür gebastelt. Das Repository dazu findet ihr [hier](https://github.com/aletutto/fae-antwort-webinterface). Für die Zukunft können wir mal überlegen, ob wir irgendein leichtgewichtiges Framework dafür verwenden wollen, oder ob es tatsächlich so schon reicht und wir das Ganze nur noch ein wenig schöner aussehen lassen wollen.

### Kurze Erklärung dazu
Eine Mail wird in dem Moment verschickt, in dem eine Ausnahmesituation erstellt wird. Diese Mail enthält einen Link zur Weboberfläche, wobei der Link als URL-Parameter die ID der Nachricht enthält. Über die Buttons auf der Weboberfläche wird dann ein POST-Request durchgeführt, welcher eine Antwort auf eine Nachricht erzeugt.

Die Empfänger-Mailadresse ist aktuell noch hardcoded. Wenn ihr das mal testen wollt, dann könnt ihr diese im `AusnahmesituationController` anpassen.

aletutto committed:
Erster Versuch für das Versenden einer E-Mail an eine Kontaktperson

Jaykju commented on src/main/java/de/th/koeln/archilab/fae/faeteam3service/ausnahmesituation/AusnahmesituationController.java:
```diff
@@ -27,11 +31,25 @@ public Ausnahmesituation createAusnahmesituation(@Valid @RequestBody Ausnahmesit
         return ausnahmesituation;
     }
 
+    private void sendMailToKontaktperson(Nachricht nachricht) {
```
Sollte das nicht eher in die `EmailServiceImpl` Klasse? Und die Funktion sollte (in Zukunft) als Parameter denke ich nicht nur die Nachricht, sondern auch die Kontaktperson und die Erkrankte Person übergeben werden, damit wir (wenn wir die Daten dann auch haben) die uns übergebenen Daten verwenden können.

aletutto replied:
Da hast du vollkommen Recht, habe es gerade in die `EmailServiceImpl` Klasse gepackt.
Und bezüglich der Parameter habe ich uns mal ein TODO dran gesetzt, das wir abarbeiten können sobald uns die nötigen Daten zur Verfügung stehen.

aletutto committed:
Cleanup des AusnahmesituationController

Die zu sendene E-Mail wird vollständig in dem dafür vorgesehen Service EmailServiceImpl zusammengebaut

Sebastian-Quast commented on src/main/java/de/th/koeln/archilab/fae/faeteam3service/antwort/AntwortController.java:
```diff
@@ -3,11 +3,7 @@
-import org.springframework.web.bind.annotation.RequestBody;
-import org.springframework.web.bind.annotation.RestController;
-
+import org.springframework.web.bind.annotation.*;
```
Würde hier nicht unbedingt auf einen wildcard Import gehen

aletutto replied:
Finde ich auch nicht so gut, hat bei mir IntelliJ wohl selbstständig gemacht.
Werde es im nächsten Commit mal abändern.

S-Lutz commented on src/main/java/de/th/koeln/archilab/fae/faeteam3service/ausnahmesituation/services/EmailServiceImpl.java:
```diff
@@ -0,0 +1,44 @@
+        text+= "Klicke auf den folgenden Link um uns mitzuteilen, ob du Maria helfen kannst:\n";
+        text += "https://fae.aletutto.de/?nachricht=" + nachricht.getNachrichtId();
+
+        return text;
```
Hier vielleicht eher sowas (?):

return String.join(String.join(System.getProperty("line.separator"),
                "Hallo Fritz,",
                "Maria benötgt gerade deine Hilfe!",
                "Klicke auf den folgenden Link um uns mitzuteilen, ob du Maria helfen kannst: ",
                "https://fae.aletutto.de/?nachricht=" + nachricht.getNachrichtId()));

aletutto replied:
Macht das Ganze besser lesbar, habs gerade eingebaut!

S-Lutz commented on src/main/java/de/th/koeln/archilab/fae/faeteam3service/antwort/AntwortController.java:
```diff
@@ -3,11 +3,7 @@
-import org.springframework.web.bind.annotation.RequestBody;
-import org.springframework.web.bind.annotation.RestController;
-
+import org.springframework.web.bind.annotation.*;
```
Würde hier nicht unbedingt auf einen wildcard Import gehen

aletutto replied:
Gefällt mir auch nicht so, das hat IntelliJ anscheinend automatisch gemacht, habs wieder auf die einzelnen Imports geändert

S-Lutz commented on src/main/resources/application.properties:
```diff
@@ -1 +1,8 @@
 server.servlet.context-path=/team3
+
+spring.mail.host=smtp.gmail.com
+spring.mail.port=587
+spring.mail.username=fae.nachrichtensystem@gmail.com
+spring.mail.password=qpkesrtalaxkpkiv
+spring.mail.properties.mail.smtp.auth=true
+spring.mail.properties.mail.smtp.starttls.enable=true
```
Wäre das ein Fall für die docker-compose?

aletutto replied:
Hast du dabei an etwas bestimmtes gedacht?
Ich habe mich an [diesem Tutorial](https://www.baeldung.com/spring-email) orientiert und dort ist es üblich die Mail Konfigurationen in die Spring Konfigurationsdatei zu packen. Lasse mich da aber auch gerne belehren falls es andere Best Practices gibt :)

S-Lutz replied:
Eventuell liege ich da falsch aber ich glaube ich habe mal gelesen, dass man Sachen wie Passwörter und so als Environment Variable zur Laufzeit reingibt, weil man die nicht in seinem Docker image haben will.

In unserem Kontext ist das wahrscheinlich nicht so wichtig und das ist vermutlich ja auch nicht dein privater Mailaccount :-D

aletutto replied:
Macht definitiv Sinn, werde mich da mal ein wenig einlesen.
Und nein, die Mail Adresse habe ich extra dafür erstellt und das Passwort ist auch "nur" ein Applikations-Passwort

S-Lutz replied:
[Docker secrets](https://docs.docker.com/engine/swarm/secrets/) könnte man sich mal anschauen.

Edit: Geht anscheinend nur mit Docker swarm

aletutto replied:
Oh okay, schade!

S-Lutz reviewed (commented):
Sieht schonmal ganz gut aus!

aletutto committed:
Kleine Änderungen durch Review von Sebastian-Quast

- Zusammenbauen der E-Mail Nachricht in der Klasse EmailServiceImpl verbessert
- Automatisch erstellte Wildcard Imports zu Single Imports im AntwortController geändert
